pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,856 @@
+            raise ValueError("invalid option for 'closed': {closed}"
+                             .format(closed=self.closed))
+        if len(self.left) != len(self.right):
+            raise ValueError('left and right must have the same length')
+        left_mask = notna(self.left)
+        right_mask = notna(self.right)
+        if not (left_mask == right_mask).all():
+            raise ValueError('missing values must be missing in the same '
+                             'location both left and right sides')
+        if not (self.left[left_mask] <= self.right[left_mask]).all():
+            raise ValueError('left side of interval must be <= right side')
+
+    # ---------
+    # Interface
+    # ---------
+    def __iter__(self):
+        return iter(self.values)
+
+    def __len__(self):
+        return len(self.left)
+
+    def __getitem__(self, value):
+        mask = self.isna()[value]
+        if is_scalar(mask) and mask:
+            return self._fill_value
+
+        left = self.left[value]
+        right = self.right[value]
+
+        # scalar
+        if not isinstance(left, Index):
+            return Interval(left, right, self.closed)
+
+        return self._shallow_copy(left, right)
+
+    def __setitem__(self, key, value):
+        if not (is_interval_dtype(value) or isinstance(value, Interval)):
+            msg = "'value' should be an interval type, got {} instead."
+            raise TypeError(msg.format(type(value)))
+
+        if value.closed != self.closed:
+            msg = "'value.closed' ({}) does not match {}."
+            raise ValueError(value.closed, self.closed)
+
+        # Need to ensure that left and right are updated atomically, so we're
+        # forced to copy, update the copy, and swap in the new values.
+        left = self.left.copy(deep=True)
+        right = self.right.copy(deep=True)
+
+        left.values[key] = value.left
+        right.values[key] = value.right
+
+        self._left = left
+        self._right = right
+
+    def fillna(self, value=None, method=None, limit=None):
+        if method is not None:
+            raise TypeError('Filling by method is not supported for '
+                            'IntervalArray.')
+        if limit is not None:
+            raise TypeError('limit is not supported for IntervalArray.')
+
+        if not isinstance(value, Interval):
+            msg = ("Interval.fillna only supports filling with scalar "
+                   "'value'. Got a '{}' instead of a 'pandas.Interval'."
+                   .format(type(value).__name__))
+            raise TypeError(msg)
+
+        value = getattr(value, '_values', value)
+
+        left = self.left.fillna(value=value.left)
+        right = self.right.fillna(value=value.right)
+        return self._shallow_copy(left, right)
+
+    @property
+    def dtype(self):
+        return IntervalDtype.construct_from_string(str(self.left.dtype))
+
+    def astype(self, dtype, copy=True):
```

Comment:
doc-string.  also these should have an ``errors=`` parameters. We do everywhere else (should be pushed to EA generally)

Reply:
I am not sure it is needed to add the `errors` keyword to our arrays (Categorical also does not have this, and none of the indexes have it). For me it would be fine to keep this as an extra feature on `Series.astype`